jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -352,117 +353,259 @@ public void abort(Throwable failure)
+                consumer.accept(this);
+                postSend();
+            }
+            else
+            {
+                callback.failed(failure);
+            }
+        }
+
+        private Throwable preSend(Callback callback, boolean commit)
+        {
+            synchronized (this)
+            {
+                switch (_state)
+                {
+                    case IDLE:
+                    {
+                        _state = State.PRE_SEND;
+                        _callback = callback;
+                        _commit = commit;
+                        return null;
+                    }
+                    default:
+                    {
+                        Throwable failure = _failure;
+                        if (failure == null)
+                            failure = new IllegalStateException("Invalid transport state: " + _state);
+                        return failure;
+                    }
+                }
+            }
+        }
+
+        private void postSend()
+        {
+            Callback callback;
+            boolean commit;
             Throwable failure;
             synchronized (this)
             {
-                state = this.state;
-                failure = this.failure;
-                if (state == State.IDLE)
+                switch (_state)
                 {
-                    this.state = State.WRITING;
-                    this.callback = callback;
-                    this.commit = commit;
-                    return true;
+                    case PRE_SEND:
+                    {
+                        _state = State.POST_SEND;
+                        return;
+                    }
+                    case SUCCEED:
+                    {
+                        callback = _callback;
+                        commit = _commit;
+                        failure = null;
+                        reset();
+                        break;
+                    }
+                    case FAIL:
+                    {
+                        _state = State.FAILED;
+                        callback = _callback;
+                        commit = _commit;
+                        failure = _failure;
+                        break;
+                    }
+                    default:
+                    {
+                        _state = State.FAILED;
+                        _failure = new IllegalStateException("Invalid transport state: " + _state);
+                        callback = _callback;
+                        commit = _commit;
+                        failure = _failure;
+                        break;
+                    }
                 }
             }
             if (failure == null)
-                failure = new IllegalStateException("Invalid transport state: " + state);
-            callback.failed(failure);
-            return false;
+                succeed(callback, commit);
+            else
+                fail(callback, commit, failure);
```

Comment:
what if callback is null?  I think it is an ISE, so that should be explicitly thrown